src/wallet/wallet.cpp

```diff
@@ -2534,8 +2534,17 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
                     return false;
                 }
 
-                if (nFeeRet >= nFeeNeeded)
+                if (nFeeRet >= nFeeNeeded) {
+                    // Reduce fee to only the needed amount if we have change output to increase
+                    // Do not do this if fee was paid by payee
+                    if (nFeeRet > nFeeNeeded && nChangePosInOut != -1 && nSubtractFeeFromAmount == 0) {
```

Comment:
I guess there's an edge case left here, where if there's no change output and the fee difference is very large, then we won't reduce the fee?

It might be nice to refactor the change-adding logic so that we could add change in that scenario, though I don't want to hold up this PR which is already a strict improvement, do you think it's worth adding a comment that points this out though for future reference?